scipy/optimize/_gensa.py

```diff
@@ -0,0 +1,601 @@
+"""
+gensa: A generalized simulated annealing global optimization algorithm
+"""
+from __future__ import division, print_function, absolute_import
+
+import numpy as np
+import time
+from scipy.optimize import OptimizeResult
+from scipy.optimize import _lbfgsb
+from scipy.special import gammaln
+
+__all__ = ['gensa']
+
+KSPRING = 1.e8
+BIG_VALUE = 1.e13
+MAX_REINIT_COUNT = 100
+
+
+class GenSARunnerException(Exception):
+    pass
+
+
+class GenSARunner(object):
+    """This class implements the core of the gensa algorithm.
+
+    x0 : ndarray
+        The starting coordinates.
+    fun : callable
+        The objective function
+    lower: ndarray
+        The lower bounds of the search domain
+    upper: ndarray
+        The upper bounds of the search domain
+    fargs: dict
+        Additional arguments to pass to the objective function
+
+    """
+
+    def __init__(self, fun, x0, lower, upper, fargs=()):
+        self._lower = np.array(lower)
+        self._upper = np.array(upper)
+        assert(len(lower) == len(
+            upper)),'Lower and upper bounds are inconsistent'
+        if x0 is None:
+            x0 = self._lower + np.random.rand(len(self._lower)) * (
+                    self._upper - self._lower)
+        else:
+            x0 = np.array(x0)
+        assert(len(self._lower) == len(
```

Comment:
`assert` can get optimised out of code. Use:

    if self._lower.size != x0.size:
        raise ValueError("Lower bounds size does not match x0")

Do the same for the upper bounds. However, the error message has to make sense for the person using the `gensa` function. There they are asked to supply `(min, max)` pairs, so the error message for someone calling `gensa` should be "the bounds tuple doesn't match x0", or something along those lines.